src/nlayout.rs

```diff
@@ -0,0 +1,57 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2022
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use pyo3::prelude::*;
+
+use hashbrown::HashMap;
+
+#[pyclass(module = "stoachstic_swap")]
+#[pyo3(text_signature = "(/")]
+#[derive(Clone, Debug)]
+pub struct NLayout {
+    pub logic_to_phys: Vec<usize>,
+    pub phys_to_logic: Vec<usize>,
+}
+
+impl NLayout {
+    pub fn swap(&mut self, idx1: usize, idx2: usize) {
+        self.phys_to_logic.swap(idx1, idx2);
+        self.logic_to_phys[self.phys_to_logic[idx1]] = idx1;
+        self.logic_to_phys[self.phys_to_logic[idx2]] = idx2;
+    }
+}
+
+#[pymethods]
+impl NLayout {
+    #[new]
+    fn new(
+        qubit_indices: HashMap<usize, usize>,
+        logical_qubits: usize,
+        physical_qubits: usize,
+    ) -> Self {
+        let mut res = NLayout {
+            logic_to_phys: vec![0; logical_qubits],
+            phys_to_logic: vec![0; physical_qubits],
+        };
+        for (key, value) in qubit_indices {
+            res.logic_to_phys[key] = value;
+            res.phys_to_logic[value] = key;
+        }
+        res
+    }
+
+    fn layout_mapping(&self) -> Vec<[usize; 2]> {
+        (0..self.logic_to_phys.len())
+            .map(|i| [self.logic_to_phys[i], self.phys_to_logic[i]])
+            .collect()
```

Comment:
This does not seem right to me (and I can imagine this is one reason of the failing tests) since the first element of `[usize; 2]` array is an index of a physical qubit but later on in lines 

https://github.com/Qiskit/qiskit-terra/blob/d72d6c5e3ba90409eeddf920673fae24b325df8a/qiskit/transpiler/passes/routing/stochastic_swap.py#L216-L218

you're using it to access a virtual qubit.

I'd go with something like:

```suggestion
        (0..self.logic_to_phys.len())
            .map(|i| [i, self.logic_to_phys[i]])
            .collect()
```

Reply:
Nice catch! I'm not sure what I was thinking here but you're clearly correct what I had was incorrect. This should definitely fix a chunk of the test failures (including the one I started debugging earlier this morning). I think there are still some other failure conditions we'll need to track down (there is also the test which is getting stuck which will be tricky to track down too).

Reply:
I was actually wrong this fixes all the errors we were facing before. The only test failures after applying this seem to be caused by rng differences (see #1794) so we probably just need to update the expected results in those cases and this should fix all the tests.